File: app/src/main/java/com/absinthe/kage/server/KageServer.java
package com.absinthe.kage.server;

import android.os.Environment;
import android.util.Log;

import com.absinthe.kage.protocol.Config;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

import fi.iki.elonen.NanoHTTPD;

public class KageServer extends NanoHTTPD {

    private static final String TAG = KageServer.class.getSimpleName();
    private static final String ROOT_DIR = "/";

    public KageServer() {
        super(Config.HTTP_SERVER_PORT);
    }

    @Override
    public Response serve(IHTTPSession session) {
        String filepath = session.getUri().trim();
        File rootDir = Environment.getExternalStorageDirectory();
        File[] filesList;
        Log.d(TAG, "Session Uri = " + filepath);
        Log.d(TAG, "Environment.getExternalStorageDirectory() = " + Environment.getExternalStorageDirectory().getAbsolutePath());

        if (filepath.trim().isEmpty() || filepath.trim().equals(ROOT_DIR)) {
            filepath = rootDir.getAbsolutePath();
        }
        filesList = new File(filepath).listFiles();

        if (filesList != null) {
            return responsePage(filesList);
        } else {
            return responseFile(filepath);
        }
    }

    //对于请求目录的，返回文件列表
    public Response responsePage(File[] filesList) {
        StringBuilder builder = new StringBuilder();
        builder.append("<!DOCTYPER html><html><body>");
        builder.append("<ol>");
        for (File detailsOfFiles : filesList) {
            builder.append("<a href=\"").append(detailsOfFiles.getAbsolutePath()).append("\" alt = \"\">").append(detailsOfFiles.getAbsolutePath()).append("</a><br>");
        }
        builder.append("</ol>");
        builder.append("</body></html>\n");
        //回送应答
        return newFixedLengthResponse(String.valueOf(builder));
    }

    //对于请求文件的，返回下载的文件
    public Response responseFile(String uri) {
        try {
            //文件输入流
            FileInputStream fis = new FileInputStream(uri);
            // 返回OK，同时传送文件，为了安全这里应该再加一个处理，即判断这个文件是否是我们所分享的文件，避免客户端访问了其他个人文件
            return newFixedLengthResponse(Response.Status.OK, "application/octet-stream", fis, fis.available());
        } catch (IOException e) {
            e.printStackTrace();
        }
        return response404(uri);
    }

    //页面不存在，或者文件不存在时
    public Response response404(String url) {
        StringBuilder builder = new StringBuilder();
        builder.append("<!DOCTYPE html><html>body>");
        builder.append("Sorry,Can't Found").append(url).append(" !");
        builder.append("</body></html>\n");
        return newFixedLengthResponse(builder.toString());
    }
}

File: app/src/main/kotlin/com/absinthe/kage/device/server/KageServer.kt
package com.absinthe.kage.device.server

import android.os.Environment
import com.absinthe.kage.connect.protocol.Config
import fi.iki.elonen.NanoHTTPD
import timber.log.Timber
import java.io.File
import java.io.FileInputStream
import java.io.IOException

class KageServer : NanoHTTPD(Config.HTTP_SERVER_PORT) {

    override fun serve(session: IHTTPSession): Response {
        val rootDir = Environment.getExternalStorageDirectory()
        val filesList: Array<File>?
        val userAgent = session.headers["user-agent"] ?: ""
        var filepath = session.uri.trim { it <= ' ' }

        Timber.d("Session Uri = $filepath")
        Timber.d("header = $userAgent")

        if (userAgent.contains("Mozilla/5.0")) {
            return responseNoAccess()
        }

        if (filepath.isEmpty() || filepath == ROOT_DIR) {
            filepath = rootDir.absolutePath
        }
        filesList = File(filepath).listFiles()
        return filesList?.let { responsePage(it) } ?: responseFile(filepath)
    }

    /**
     * Return files list if request a directory
     *
     * @param filesList Files list
     */
    private fun responsePage(filesList: Array<File>): Response {
        val builder = StringBuilder().apply {
            append("<!DOCTYPER html><html><body>")
            append("<ol>")

            for (detailsOfFiles in filesList) {
                append("<a href=\"")
                append(detailsOfFiles.absolutePath).append("\" alt = \"\">")
                append(detailsOfFiles.absolutePath).append("</a><br>")
            }

            append("</ol>")
            append("</body></html>\n")
        }

        //Response
        return newFixedLengthResponse(builder.toString())
    }

    /**
     * Return itself if request a file
     *
     * @param uri URI of the file
     */
    private fun responseFile(uri: String): Response {
        try {
            val fis = FileInputStream(uri)
            // Response OK and transfer the file
            return newFixedLengthResponse(Response.Status.OK,
                    "application/octet-stream", fis, fis.available().toLong())
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return responseNotExist(uri)
    }

    /**
     * Return Error info if request an invalid path
     *
     * @param url Invalid URL
     */
    private fun responseNotExist(url: String): Response {
        val builder = StringBuilder().apply {
            append("<!DOCTYPE html><html><body>")
            append("Sorry,Can't Found").append(url).append(" !")
            append("</body></html>\n")
        }

        return newFixedLengthResponse(builder.toString())
    }

    /**
     * Refuse to access the server if it is not a client
     */
    private fun responseNoAccess(): Response {
        val builder = StringBuilder().apply {
            append("<!DOCTYPE html><html><body>")
            append("No Access!")
            append("</body></html>\n")
        }

        return newFixedLengthResponse(builder.toString())
    }

    companion object {
        private const val ROOT_DIR = "/"
    }
}